server/src/test/java/org/elasticsearch/action/admin/cluster/settings/ClusterGetSettingsTests.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.List;
+
+import static org.mockito.Mockito.mock;
+
+public class ClusterGetSettingsTests extends ESTestCase {
+
+    public void testRequestConstruction() {
+        final Settings persistentSettings = Settings.builder()
+            .put("persistent.foo.filtered", "bar")
+            .put("persistent.foo.non_filtered", "baz")
+            .build();
+
+        final Settings transientSettings = Settings.builder()
+            .put("transient.foo.filtered", "bar")
+            .put("transient.foo.non_filtered", "baz")
+            .build();
+
+        ClusterGetSettingsAction.Response response = new ClusterGetSettingsAction.Response(persistentSettings, transientSettings, null);
+
+        assertEquals(persistentSettings, response.persistentSettings());
+        assertEquals(transientSettings, response.transientSettings());
+        assertEquals(Settings.EMPTY, response.settings());
+    }
+
+    public void testTransportFilters() throws Exception {
+        final SettingsFilter filter = new SettingsFilter(List.of("persistent.foo.filtered", "transient.foo.filtered"));
+
+        TransportClusterGetSettingsAction action = new TransportClusterGetSettingsAction(
+            mock(TransportService.class),
+            mock(ClusterService.class),
+            mock(ThreadPool.class),
+            filter,
+            mock(ActionFilters.class),
+            mock(IndexNameExpressionResolver.class)
+        );
+
+        final Settings persistentSettings = Settings.builder()
+            .put("persistent.foo.filtered", "bar")
+            .put("persistent.foo.non_filtered", "baz")
+            .build();
+
+        final Settings transientSettings = Settings.builder()
+            .put("transient.foo.filtered", "bar")
+            .put("transient.foo.non_filtered", "baz")
+            .build();
+
+        final Metadata metadata = Metadata.builder().persistentSettings(persistentSettings).transientSettings(transientSettings).build();
+        final ClusterState clusterState = ClusterState.builder(ClusterState.EMPTY_STATE).metadata(metadata).build();
+
+        action.masterOperation(null, null, clusterState, new ActionListener<>() {
+            @Override
+            public void onResponse(ClusterGetSettingsAction.Response response) {
+                assertFalse(response.persistentSettings().hasValue("persistent.foo.filtered"));
+                assertTrue(response.persistentSettings().hasValue("persistent.foo.non_filtered"));
+
+                assertFalse(response.transientSettings().hasValue("transient.foo.filtered"));
+                assertTrue(response.transientSettings().hasValue("transient.foo.non_filtered"));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail("Shouldn't reach here");
+            }
+        });
+    }
```

Comment:
The trouble with this pattern is that it would pass even if the listener was not completed. There's various ways to avoid that, for instance using a future:

```suggestion
        final PlainActionFuture<ClusterGetSettingsAction.Response> future = new PlainActionFuture<>();
        action.masterOperation(null, null, clusterState, future);
        assertTrue(future.isDone());

        final ClusterGetSettingsAction.Response response = future.get();

        assertFalse(response.persistentSettings().hasValue("persistent.foo.filtered"));
        assertTrue(response.persistentSettings().hasValue("persistent.foo.non_filtered"));

        assertFalse(response.transientSettings().hasValue("transient.foo.filtered"));
        assertTrue(response.transientSettings().hasValue("transient.foo.non_filtered"));
```